container/lxd/container.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package lxd
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/network"
+	"github.com/juju/utils/arch"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+const UserNamespacePrefix = "user."
+
+// ContainerSpec represents the data required to create a new container.
+type ContainerSpec struct {
+	Name     string
+	Image    SourcedImage
+	Devices  map[string]device
+	Config   map[string]string
+	Profiles []string
+}
+
+// Normalise config ensures that configuration keys are prefixed with the
+// "user" namespace where appropriate.
+func (c *ContainerSpec) NormaliseConfig() {
+	newConfig := make(map[string]string, len(c.Config))
+	for k, v := range c.Config {
+		if strings.HasPrefix(k, UserNamespacePrefix) || strings.HasPrefix(k, "limits.") || k == "boot.autostart" {
+			newConfig[k] = v
+			continue
+		}
+		newConfig[UserNamespacePrefix+k] = v
+	}
+	c.Config = newConfig
+}
+
+// Container extends the upstream LXD container type.
+type Container struct {
+	api.Container
+}
+
+// Metadata returns the value from container config for the input key.
+// Such values are stored with the "user" namespace prefix.
+func (c *Container) Metadata(key string) string {
+	return c.Config[UserNamespacePrefix+key]
+}
+
+// Arch returns the architecture of the container.
+func (c *Container) Arch() string {
+	return arch.NormaliseArch(c.Architecture)
+}
+
+// CPUs returns the configured limit for number of container CPU cores.
+// If unset, zero is returned.
+func (c *Container) CPUs() uint {
+	var cores uint = 0
+	if v := c.Config["limits.cpu"]; v != "" {
+		fmt.Sscanf(v, "%d", &cores)
```

Comment:
We should probably care about the error from `fmt.Sscanf` here, or at least if there is an error just return 0?